### Examen T4 - Manipulación de Dataframes

**Pregunta 1:** El fichero cotizacion.csv contiene las cotizaciones de las empresas del IBEX35 con las siguientes columnas: 

* Nombre (nombre de la empresa) 
* Final (precio de la acción al cierre de bolsa) 
* Máximo (precio máximo de la acción durante la jornada) 
* Mínimo (precio mínimo de la acción durante la jornada) 
* Volumen (Volumen al cierre de bolsa)
* Efectivo (capitalización al cierre en miles de euros)

Filtrar en un nuevo dataset aquellos registros cuyo promedio entre el "Máximo" y el "Mínimo" sea mayor al "Final" e imprimir el número de registros.



In [1]:
import pandas as pd
import numpy as np

In [147]:
cotizaciones = pd.read_csv("cotizacion.csv", sep=";")
cotizaciones.head()

Nombre  Final Máximo Mínimo  Volumen  Efectivo
0   ACCIONA  95,95  96,75   94,4   84.962   8166,11
1  ACERINOX  86,68  86,72  84,68   885.99   7633,81
2       ACS  37,28  37,66  37,22  655.982  24517,29
3      AENA  167,1  167,5  166,1  133.738  22336,10
4   AMADEUS  69,96  70,64  69,56  640.476  44795,47

In [148]:
#columnas_a_convertir = ["Final","Máximo","Mínimo","Volumen","Efectivo"]
#cotizaciones["columnas_a_convertir"] = cotizaciones["columnas_a_convertir"].str.replace(',', '.').astype(float)
cotizaciones["Final"] = cotizaciones["Final"].str.replace(',', '.')
cotizaciones["Final"] = pd.to_numeric(cotizaciones["Final"])
cotizaciones["Máximo"] = cotizaciones["Máximo"].str.replace(',', '.').astype(float)
cotizaciones["Mínimo"] = cotizaciones["Mínimo"].str.replace(',', '.').astype(float)
cotizaciones["Volumen"] = cotizaciones["Volumen"].str.replace('.', '')
cotizaciones["Volumen"] = pd.to_numeric(cotizaciones["Volumen"])
cotizaciones["Efectivo"] = cotizaciones["Efectivo"].str.replace(',', '.').astype(float)
cotizaciones.head()

Nombre   Final  Máximo  Mínimo  Volumen  Efectivo
0   ACCIONA   95.95   96.75   94.40    84962   8166.11
1  ACERINOX   86.68   86.72   84.68    88599   7633.81
2       ACS   37.28   37.66   37.22   655982  24517.29
3      AENA  167.10  167.50  166.10   133738  22336.10
4   AMADEUS   69.96   70.64   69.56   640476  44795.47

In [149]:
tipo3=cotizaciones["Volumen"].dtypes
print(tipo3)

int64


In [150]:
cotizaciones["Promedio"]=(cotizaciones["Máximo"]+cotizaciones["Mínimo"])/2
cotizaciones

Nombre     Final    Máximo    Mínimo   Volumen   Efectivo   Promedio
0        ACCIONA   95.9500   96.7500   94.4000     84962    8166.11   95.57500
1       ACERINOX   86.6800   86.7200   84.6800     88599    7633.81   85.70000
2            ACS   37.2800   37.6600   37.2200    655982   24517.29   37.44000
3           AENA  167.1000  167.5000  166.1000    133738   22336.10  166.80000
4        AMADEUS   69.9600   70.6400   69.5600    640476   44795.47   70.10000
5    ARCELORMIT.   14.1520   14.3180   14.0800    702841    9989.65   14.19900
6    BA.SABADELL    1.0165    1.0330    1.0165  12358512   12637.65    1.02475
7   BA.SANTANDER    4.0200    4.0675    4.0100  36129692  145765.44    4.03875
8         BANKIA    2.2490    2.2620    2.2240   3024454    6803.54    2.24300
9      BANKINTER    6.5600    6.6040    6.5220   1375976    9030.20    6.56300
10          BBVA    5.0240    5.0500    4.9690  26447612  132560.15    5.00950
11     CAIXABANK    2.7630    2.8060    2.7620  13031097   36239.22    2.78400
12       CELLNEX   30.1700   30.7000   30.1700    857237   25965.29   30.43500
13  CIE AUTOMOT.   23.4400   23.7400   23.4000     99506    2343.09   23.57000
14        ENAGAS   25.6000   25.6600   25.0700     84254   21524.77   25.36500
15          ENCE    3.8820    3.9740    3.8460   2099698    8219.83    3.91000
16        ENDESA   22.9500   23.0300   22.6200   1236799   28233.46   22.82500
17     FERROVIAL   21.6100   21.7200   21.3800   2065931   44656.05   21.55000
18  GRIFOLS CL.A   23.3000   23.6000   23.2700    719503   16787.55   23.43500
19           IAG    5.3000    5.4600    5.2920   2155081   11561.95    5.37600
20     IBERDROLA    8.4500    8.4500    8.2620  15637958  131426.05    8.35600
21       INDITEX   24.9300   25.1900   24.8700   2600144   65012.92   25.03000
22       INDRA A    9.5600    9.6850    9.5350     39206    3755.57    9.61000
23  INM.COLONIAL    9.9150    9.9150    9.7450    671753    6640.38    9.83000
24        MAPFRE    2.6540    2.6540    2.6200   3045972    8063.13    2.63700
25      MEDIASET    6.4460    6.5920    6.4460   2647557   17277.04    6.51900
26  MELIA HOTELS    8.1400    8.1900    8.0200     48021    3914.93    8.10500
27        MERLIN   12.0800   12.0800   11.8300    777543    9357.77   11.95500
28       NATURGY   26.7000   26.7700   26.5300   2100737   55986.54   26.65000
29        R.E.C.   19.7050   19.8750   19.6750   1246171   24585.33   19.77500
30        REPSOL   14.5150   14.6050   14.4600   5046648   73333.78   14.53250
31  SIEMENS GAME   14.3300   14.3900   14.1550   1054847   15084.56   14.27250
32  TEC.REUNIDAS   23.0000   23.2000   22.7800      1221    2807.19   22.99000
33    TELEFONICA    7.2040    7.2560    7.1850   9913249   71564.99    7.22050
34      VISCOFAN   48.4600   49.1200   48.4600      6771    3295.43   48.79000

Filtrar en un nuevo dataset aquellos registros cuyo promedio entre el "Máximo" y el "Mínimo" sea mayor al "Final" e imprimir el número de registros.

In [151]:
df_cotizaciones = cotizaciones[cotizaciones["Promedio"]>cotizaciones["Final"]]
df_cotizaciones

Nombre    Final   Máximo   Mínimo   Volumen   Efectivo  Promedio
2            ACS  37.2800  37.6600  37.2200    655982   24517.29  37.44000
4        AMADEUS  69.9600  70.6400  69.5600    640476   44795.47  70.10000
5    ARCELORMIT.  14.1520  14.3180  14.0800    702841    9989.65  14.19900
6    BA.SABADELL   1.0165   1.0330   1.0165  12358512   12637.65   1.02475
7   BA.SANTANDER   4.0200   4.0675   4.0100  36129692  145765.44   4.03875
9      BANKINTER   6.5600   6.6040   6.5220   1375976    9030.20   6.56300
11     CAIXABANK   2.7630   2.8060   2.7620  13031097   36239.22   2.78400
12       CELLNEX  30.1700  30.7000  30.1700    857237   25965.29  30.43500
13  CIE AUTOMOT.  23.4400  23.7400  23.4000     99506    2343.09  23.57000
15          ENCE   3.8820   3.9740   3.8460   2099698    8219.83   3.91000
18  GRIFOLS CL.A  23.3000  23.6000  23.2700    719503   16787.55  23.43500
19           IAG   5.3000   5.4600   5.2920   2155081   11561.95   5.37600
21       INDITEX  24.9300  25.1900  24.8700   2600144   65012.92  25.03000
22       INDRA A   9.5600   9.6850   9.5350     39206    3755.57   9.61000
25      MEDIASET   6.4460   6.5920   6.4460   2647557   17277.04   6.51900
29        R.E.C.  19.7050  19.8750  19.6750   1246171   24585.33  19.77500
30        REPSOL  14.5150  14.6050  14.4600   5046648   73333.78  14.53250
33    TELEFONICA   7.2040   7.2560   7.1850   9913249   71564.99   7.22050
34      VISCOFAN  48.4600  49.1200  48.4600      6771    3295.43  48.79000

**Pregunta 2:** Dado el dataset anterior filtrar el "Efectivo" mediante el método between los valores entre 11561.95 y 33000. No incluir los limites.

In [152]:
df_efectivo = df_cotizaciones[df_cotizaciones['Efectivo'].between(11561.95, 33000, inclusive='neither')]
df_efectivo

Nombre    Final  Máximo   Mínimo   Volumen  Efectivo  Promedio
2            ACS  37.2800  37.660  37.2200    655982  24517.29  37.44000
6    BA.SABADELL   1.0165   1.033   1.0165  12358512  12637.65   1.02475
12       CELLNEX  30.1700  30.700  30.1700    857237  25965.29  30.43500
18  GRIFOLS CL.A  23.3000  23.600  23.2700    719503  16787.55  23.43500
25      MEDIASET   6.4460   6.592   6.4460   2647557  17277.04   6.51900
29        R.E.C.  19.7050  19.875  19.6750   1246171  24585.33  19.77500

**Pregunta 3:** Dado el dataset "estudiantes.csv":

* Agregar la columna promedio que contenga el promedio de las 3 notas.
* Utiliza el método query() para filtrar el DataFrame y seleccionar solo las filas donde el nombre del estudiante comienza con una letra específica (por ejemplo, 'A') y donde haya obtenido una calificación superior al promedio en al menos dos asignaturas.
* Almacena el resultado filtrado en un nuevo DataFrame llamado calificaciones_filtradas.
* Muestra el DataFrame calificaciones_filtradas.

In [153]:
df_estudiantes = pd.read_csv("estudiantes.csv", sep="-")
df_estudiantes

Unnamed: 0     Nombre  Asignatura1  Asignatura2  Asignatura3
0             0   Artesino           68           86           62
1             1     Salome           63           74           92
2             2     Salome           64           63           71
3             3  NievaNuma           82           73           71
4             4    Corsino           76           84           89
..          ...        ...          ...          ...          ...
145         145   Bautista           93           78           93
146         146      Seida           74           67           93
147         147    Macario           77           89           72
148         148    Balbino           92           94           83
149         149       Eloa           85           60           72

[150 rows x 5 columns]

In [154]:
nombre_columna = df_estudiantes.columns[0]
df_estudiantes = df_estudiantes.drop(columns=nombre_columna)
df_estudiantes

Nombre  Asignatura1  Asignatura2  Asignatura3
0     Artesino           68           86           62
1       Salome           63           74           92
2       Salome           64           63           71
3    NievaNuma           82           73           71
4      Corsino           76           84           89
..         ...          ...          ...          ...
145   Bautista           93           78           93
146      Seida           74           67           93
147    Macario           77           89           72
148    Balbino           92           94           83
149       Eloa           85           60           72

[150 rows x 4 columns]

* Agregar la columna promedio que contenga el promedio de las 3 notas.

In [155]:
df_estudiantes['Promedio']=((df_estudiantes["Asignatura1"]+df_estudiantes["Asignatura2"]+df_estudiantes["Asignatura3"])/3).round(2)
df_estudiantes

Nombre  Asignatura1  Asignatura2  Asignatura3  Promedio
0     Artesino           68           86           62     72.00
1       Salome           63           74           92     76.33
2       Salome           64           63           71     66.00
3    NievaNuma           82           73           71     75.33
4      Corsino           76           84           89     83.00
..         ...          ...          ...          ...       ...
145   Bautista           93           78           93     88.00
146      Seida           74           67           93     78.00
147    Macario           77           89           72     79.33
148    Balbino           92           94           83     89.67
149       Eloa           85           60           72     72.33

[150 rows x 5 columns]

* Utiliza el método query() para filtrar el DataFrame y seleccionar solo las filas donde el nombre del estudiante comienza con una letra específica (por ejemplo, 'A') y donde haya obtenido una calificación superior al promedio en al menos dos asignaturas.

In [156]:
filtro_nombre = "Nombre.str.startswith('A')"
filtro_asignaturas = "(Asignatura1 > Promedio) & (Asignatura3 > Promedio)"

df_filtrado=df_estudiantes.query(f"{filtro_nombre} and {filtro_asignaturas}")
df_filtrado


Nombre  Asignatura1  Asignatura2  Asignatura3  Promedio
69       Artesino           77           67           80     74.67
72         Amable           96           83           90     89.67
87       Antonino           99           84           96     93.00
112        Aniano           94           86           98     92.67
121  AnharaArlena           93           72           92     85.67
127         Audaz           84           66           80     76.67
133      Artesino           95           68           85     82.67

**Pregunta 4:** Utilizando el dataset anterior categorizar las notas de la asignatura 2 en 10 categorias e imprimir los bins.

In [157]:
categorias, bins = pd.cut(df_filtrado['Asignatura2'],bins=10,retbins=True)
categorias


69     (65.98, 68.0]
72      (82.0, 84.0]
87      (82.0, 84.0]
112     (84.0, 86.0]
121     (70.0, 72.0]
127    (65.98, 68.0]
133    (65.98, 68.0]
Name: Asignatura2, dtype: category
Categories (10, interval[float64, right]): [(65.98, 68.0] < (68.0, 70.0] < (70.0, 72.0] < (72.0, 74.0] ... (78.0, 80.0] < (80.0, 82.0] < (82.0, 84.0] < (84.0, 86.0]]

In [158]:
bins

array([65.98, 68.  , 70.  , 72.  , 74.  , 76.  , 78.  , 80.  , 82.  ,
       84.  , 86.  ])

**Pregunta 5:** Utilizando el dataset anterior añadir la columna "Categoria" que se cree a partir de la columna "Promedio" con las siguientes categorias:

* 0-70 >> 'Necesita mejorar'
* 70-80 >> 'Aceptable'
* 80-90 >> 'Bueno'
* 90-100 >>'Excelente'

A continuación:

* Imprimir cuantos alumnos tienen una nota Aceptable
* Imprimir cuantos alumnos tienen una nota Excelente


In [159]:
categoria = [0,70,80,90,100]
labels_personalizados = ['Necesita mejorar', 'Aceptable', 'Bueno', 'Excelente']

df_filtrado_copia = df_filtrado.copy() #Salía advertencia de no saber si esta trabajando con dataframe original o no. Se crea una copia y se trabaja con ésta para evitar la advertencia.

df_filtrado_copia['Categoria']=pd.cut(df_filtrado['Promedio'],bins=categoria, labels=labels_personalizados)
df_filtrado_copia

Nombre  Asignatura1  Asignatura2  Asignatura3  Promedio  Categoria
69       Artesino           77           67           80     74.67  Aceptable
72         Amable           96           83           90     89.67      Bueno
87       Antonino           99           84           96     93.00  Excelente
112        Aniano           94           86           98     92.67  Excelente
121  AnharaArlena           93           72           92     85.67      Bueno
127         Audaz           84           66           80     76.67  Aceptable
133      Artesino           95           68           85     82.67      Bueno

In [160]:
excelentes_aceptables = df_filtrado_copia.groupby('Categoria', observed=False).size().reset_index(name='Total')
conteo=excelentes_aceptables[excelentes_aceptables['Categoria'].isin(['Aceptable', 'Excelente'])]
conteo

Categoria  Total
1  Aceptable      2
3  Excelente      2

**Pregunta 6:** En un conjunto de datos puede haber 10 deciles (discretizar en 10) y el decil 5 de un conjunto de datos es igual a la mediana. Utilizando la columna "Asignatura2" comprobar mediante el método "qcut" si correcta esta afirmación.

In [161]:
df_filtrado_copia['Discretizar_en_10'] = pd.qcut(df_filtrado_copia['Asignatura2'], q=10, labels=['Q1','Q2','Q3','Q4','Q5','Q6','Q7','Q8','Q9','Q10'])
df_filtrado_copia

Nombre  Asignatura1  Asignatura2  Asignatura3  Promedio  Categoria  \
69       Artesino           77           67           80     74.67  Aceptable   
72         Amable           96           83           90     89.67      Bueno   
87       Antonino           99           84           96     93.00  Excelente   
112        Aniano           94           86           98     92.67  Excelente   
121  AnharaArlena           93           72           92     85.67      Bueno   
127         Audaz           84           66           80     76.67  Aceptable   
133      Artesino           95           68           85     82.67      Bueno   

    Discretizar_en_10  
69                 Q2  
72                 Q7  
87                 Q9  
112               Q10  
121                Q5  
127                Q1  
133                Q4

In [162]:
mediana= df_filtrado_copia['Asignatura2'].median()
mediana

72.0

Se comprueba que que el Q5 es igual a la mediana, ambos son 72.

**Pregunta 7:** Dado los datasets de ventas de enero y febrero:

* Crea un dataframe por cada dataset.
* Añadir la columna mes a cada dataframe que contenga el valor del mes.
* Concatenar ambos datasets a travez del eje 0

Utilizando GroupBy

* Obtener la cantidad total vendida del producto A y B.
* Obtener los ingresos totales en el mes de febrero.

In [198]:
datos_enero = {
    'Producto': ['A', 'B', 'C'],
    'Cantidad': [100, 150, 200],
    'Ingresos': [1000, 2000, 3000]
}

datos_febrero = {
    'Producto': ['A', 'B', 'D'],
    'Cantidad': [120, 180, 220],
    'Ingresos': [1200, 2200, 3300]
}

In [199]:
df_datos_enero = pd.DataFrame(datos_enero)
df_datos_enero

Producto  Cantidad  Ingresos
0        A       100      1000
1        B       150      2000
2        C       200      3000

In [200]:
df_datos_febrero = pd.DataFrame(datos_febrero)
df_datos_febrero

Producto  Cantidad  Ingresos
0        A       120      1200
1        B       180      2200
2        D       220      3300

In [201]:
df_datos_enero['Mes']='Enero'
df_datos_enero


Producto  Cantidad  Ingresos    Mes
0        A       100      1000  Enero
1        B       150      2000  Enero
2        C       200      3000  Enero

In [202]:
df_datos_febrero['Mes']='Febrero'
df_datos_febrero

Producto  Cantidad  Ingresos      Mes
0        A       120      1200  Febrero
1        B       180      2200  Febrero
2        D       220      3300  Febrero

* Concatenar ambos datasets a travez del eje 0

In [203]:
df = pd.concat([df_datos_enero, df_datos_febrero], axis=0)
df

Producto  Cantidad  Ingresos      Mes
0        A       100      1000    Enero
1        B       150      2000    Enero
2        C       200      3000    Enero
0        A       120      1200  Febrero
1        B       180      2200  Febrero
2        D       220      3300  Febrero

Utilizando GroupBy

* Obtener la cantidad total vendida del producto A y B.

In [204]:
cantidad_A  = df.groupby('Producto')
for i in cantidad_A:
    print(i)


('A',   Producto  Cantidad  Ingresos      Mes
0        A       100      1000    Enero
0        A       120      1200  Febrero)
('B',   Producto  Cantidad  Ingresos      Mes
1        B       150      2000    Enero
1        B       180      2200  Febrero)
('C',   Producto  Cantidad  Ingresos    Mes
2        C       200      3000  Enero)
('D',   Producto  Cantidad  Ingresos      Mes
2        D       220      3300  Febrero)


In [205]:
cantidad_A_B  = df.groupby('Producto')['Cantidad'].sum().loc[['A','B']]
cantidad_A_B

Producto
A    220
B    330
Name: Cantidad, dtype: int64

* Obtener los ingresos totales en el mes de febrero.

In [206]:
ingresos_feb = df.groupby('Mes')['Ingresos']
for i in ingresos_feb:
    print(i)

('Enero', 0    1000
1    2000
2    3000
Name: Ingresos, dtype: int64)
('Febrero', 0    1200
1    2200
2    3300
Name: Ingresos, dtype: int64)


In [208]:
ingresos_febrero = df.groupby('Mes')['Ingresos'].sum().loc[['Febrero']]
df_p= pd.DataFrame(ingresos_febrero)
df_p

Ingresos
Mes              
Febrero      6700

**Pregunta 8:** Dado los siguientes datasets:

* Crea dos DataFrames ventas_df y clientes_df con los datos.
* Utiliza el método merge() para combinar los dos DataFrames en uno solo, utilizando la columna ID_Cliente como clave de combinación(Revisa la documentación de los parámetros).
* Personaliza el tipo de combinación para que incluya todas las ventas y todos los clientes, incluso si no hay coincidencias en ambos conjuntos de datos.
* Reemplaza los NaN del DF resultantes por "Sin valor"
* Muestra el DataFrame combinado resultante.

In [173]:
datos_ventas = {
    'ID_Cliente': [101, 102, 103, 104],
    'Producto': ['A', 'B', 'C', 'D'],
    'Cantidad': [100, 150, 200, 250],
    'Ingresos': [1000, 2000, 3000, 4000]
}

datos_clientes = {
    'ID_Cliente': [101, 102, 105],
    'Nombre': ['Ana', 'Bob', 'Elena'],
    'Correo': ['ana@example.com', 'bob@example.com', 'elena@example.com'],
    'Dirección': ['123 Calle Principal', '456 Calle Secundaria', '789 Avenida Central']
}

In [174]:
df_ventas = pd.DataFrame(datos_ventas)
df_ventas

ID_Cliente Producto  Cantidad  Ingresos
0         101        A       100      1000
1         102        B       150      2000
2         103        C       200      3000
3         104        D       250      4000

In [175]:
df_clientes = pd.DataFrame(datos_clientes)
df_clientes

ID_Cliente Nombre             Correo             Dirección
0         101    Ana    ana@example.com   123 Calle Principal
1         102    Bob    bob@example.com  456 Calle Secundaria
2         105  Elena  elena@example.com   789 Avenida Central

* Utiliza el método merge() para combinar los dos DataFrames en uno solo, utilizando la columna ID_Cliente como clave de combinación (Revisa la documentación de los parámetros).

In [176]:
df_ventas_clientes = pd.merge(df_ventas, df_clientes, left_on='ID_Cliente', right_on='ID_Cliente')
df_ventas_clientes

ID_Cliente Producto  Cantidad  Ingresos Nombre           Correo  \
0         101        A       100      1000    Ana  ana@example.com   
1         102        B       150      2000    Bob  bob@example.com   

              Dirección  
0   123 Calle Principal  
1  456 Calle Secundaria

* Personaliza el tipo de combinación para que incluya todas las ventas y todos los clientes, incluso si no hay coincidencias en ambos conjuntos de datos.

In [177]:
df_ventas_clientes = pd.merge(df_ventas, df_clientes, how='outer', on='ID_Cliente')
df_ventas_clientes

ID_Cliente Producto  Cantidad  Ingresos Nombre             Correo  \
0         101        A     100.0    1000.0    Ana    ana@example.com   
1         102        B     150.0    2000.0    Bob    bob@example.com   
2         103        C     200.0    3000.0    NaN                NaN   
3         104        D     250.0    4000.0    NaN                NaN   
4         105      NaN       NaN       NaN  Elena  elena@example.com   

              Dirección  
0   123 Calle Principal  
1  456 Calle Secundaria  
2                   NaN  
3                   NaN  
4   789 Avenida Central

* Reemplaza los NaN del DF resultantes por "Sin valor"
* Muestra el DataFrame combinado resultante.

In [178]:
df_ventas_clientes = pd.merge(df_ventas, df_clientes, how='outer', on='ID_Cliente')
df_ventas_clientes.fillna(value='Sin valor', inplace=True)
df_ventas_clientes

C:\Users\guill\AppData\Local\Temp\ipykernel_25068\3092811133.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'Sin valor' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df_ventas_clientes.fillna(value='Sin valor', inplace=True)


ID_Cliente   Producto   Cantidad   Ingresos     Nombre             Correo  \
0         101          A      100.0     1000.0        Ana    ana@example.com   
1         102          B      150.0     2000.0        Bob    bob@example.com   
2         103          C      200.0     3000.0  Sin valor          Sin valor   
3         104          D      250.0     4000.0  Sin valor          Sin valor   
4         105  Sin valor  Sin valor  Sin valor      Elena  elena@example.com   

              Dirección  
0   123 Calle Principal  
1  456 Calle Secundaria  
2             Sin valor  
3             Sin valor  
4   789 Avenida Central

**Pregunta 9:** Dado los siguientes datasets:

* Crea dos DataFrames ventas_df y clientes_df con los datos.
* Utiliza el método join() para combinar los dos DataFrames en uno solo, utilizando la columna ID_Cliente.
* Personaliza el tipo de combinación para que incluya todas las ventas, clientes y productos, incluso si no hay coincidencias en ambos conjuntos de datos.
* Reemplaza los NaN del DF resultantes por "Sin valor"
* Muestra el DataFrame combinado resultante.

In [179]:

datos_ventas = {
    'ID_Venta': [1, 2, 3, 4],
    'ID_Cliente': [101, 102, 103, 104],
    'ID_Producto': [201, 202, 203, 204],
    'Cantidad': [100, 150, 200, 250],
    'Ingresos': [1000, 2000, 3000, 4000]
}

datos_clientes = {
    'ID_Cliente': [101, 102, 105],
    'Nombre': ['Ana', 'Bob', 'Elena'],
    'Correo': ['ana@example.com', 'bob@example.com', 'elena@example.com'],
    'Dirección': ['123 Calle Principal', '456 Calle Secundaria', '789 Avenida Central']
}


In [180]:
ventas_df = pd.DataFrame(datos_ventas)
ventas_df

ID_Venta  ID_Cliente  ID_Producto  Cantidad  Ingresos
0         1         101          201       100      1000
1         2         102          202       150      2000
2         3         103          203       200      3000
3         4         104          204       250      4000

In [181]:
clientes_df = pd.DataFrame(datos_clientes)
clientes_df

ID_Cliente Nombre             Correo             Dirección
0         101    Ana    ana@example.com   123 Calle Principal
1         102    Bob    bob@example.com  456 Calle Secundaria
2         105  Elena  elena@example.com   789 Avenida Central

* Utiliza el método join() para combinar los dos DataFrames en uno solo, utilizando la columna ID_Cliente.

In [182]:
join_resultado = ventas_df.join(clientes_df, lsuffix='ID_Cliente')
join_resultado

ID_Venta  ID_ClienteID_Cliente  ID_Producto  Cantidad  Ingresos  \
0         1                   101          201       100      1000   
1         2                   102          202       150      2000   
2         3                   103          203       200      3000   
3         4                   104          204       250      4000   

   ID_Cliente Nombre             Correo             Dirección  
0       101.0    Ana    ana@example.com   123 Calle Principal  
1       102.0    Bob    bob@example.com  456 Calle Secundaria  
2       105.0  Elena  elena@example.com   789 Avenida Central  
3         NaN    NaN                NaN                   NaN

* Personaliza el tipo de combinación para que incluya todas las ventas, clientes y productos, incluso si no hay coincidencias en ambos conjuntos de datos.

In [183]:
join_todo_result= ventas_df.set_index('ID_Cliente')
join_todo_resultado = join_todo_result.join(clientes_df.set_index('ID_Cliente'), how='outer')
join_todo_resultado

ID_Venta  ID_Producto  Cantidad  Ingresos Nombre  \
ID_Cliente                                                     
101              1.0        201.0     100.0    1000.0    Ana   
102              2.0        202.0     150.0    2000.0    Bob   
103              3.0        203.0     200.0    3000.0    NaN   
104              4.0        204.0     250.0    4000.0    NaN   
105              NaN          NaN       NaN       NaN  Elena   

                       Correo             Dirección  
ID_Cliente                                           
101           ana@example.com   123 Calle Principal  
102           bob@example.com  456 Calle Secundaria  
103                       NaN                   NaN  
104                       NaN                   NaN  
105         elena@example.com   789 Avenida Central

* Reemplaza los NaN del DF resultantes por "Sin valor"
* Muestra el DataFrame combinado resultante.

In [184]:
join_todo_resultado.fillna(value='Sin valor', inplace=True)
join_todo_resultado

C:\Users\guill\AppData\Local\Temp\ipykernel_25068\2929729147.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'Sin valor' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  join_todo_resultado.fillna(value='Sin valor', inplace=True)


ID_Venta ID_Producto   Cantidad   Ingresos     Nombre  \
ID_Cliente                                                           
101               1.0       201.0      100.0     1000.0        Ana   
102               2.0       202.0      150.0     2000.0        Bob   
103               3.0       203.0      200.0     3000.0  Sin valor   
104               4.0       204.0      250.0     4000.0  Sin valor   
105         Sin valor   Sin valor  Sin valor  Sin valor      Elena   

                       Correo             Dirección  
ID_Cliente                                           
101           ana@example.com   123 Calle Principal  
102           bob@example.com  456 Calle Secundaria  
103                 Sin valor             Sin valor  
104                 Sin valor             Sin valor  
105         elena@example.com   789 Avenida Central

### Detalle dataset titanic

**survival:**	Sobreviviente	0 = No, 1 = Yes

**pclass:**	Clase de boleto	1 = 1st, 2 = 2nd, 3 = 3rd

**sex:**	Sexo	

**Age:**	Edad en años

**sibsp:**	# de hermanos/cónyuges a bordo del Titanic

**parch:**	# de padres/hijos a bordo del Titanic

**ticket:**	Numero de ticket	

**fare:**	Tarifa de pasajero

**cabin:**	Número de cabina
	
**embarked:**	Puerto de embarque	C = Cherbourg, Q = Queenstown, S = Southampton

**--------------------------------------------------------------------------------------------**

**Pregunta 10:** Dado el dataset titanic:

* Categorizar la columna edad en Infantes,Niños,Adolescentes,Jóvenes,Adultos, Adulto Mayor y SinDato.
* Crear un nuevo dataframe que contenga la edad y las categorías.
* Apartir de la dataframe anterior obtener que porcentaje de pasajeros fueron Infantes,Niños,Adolescentes,Jóvenes,Adultos, Adulto Mayor y SinDato.
* Crear un nuevo dataframe que contenga las categorias,conteo,porcentaje, como se muestra a continuación.

In [3]:
titanic_df = pd.read_csv('titanic.csv')
titanic_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
category = [0, 5, 12, 18, 30, 65, 100]
labels_titanic = ['Infantes', 'Niños', 'Adolescentes', 'Jóvenes', 'Adultos', 'Adulto Mayor']

titanic_df['Category'] = pd.cut(titanic_df['Age'], bins=category, labels=labels_titanic)

titanic_df['Category'] = titanic_df['Category'].cat.add_categories('SinDato').fillna('SinDato')
titanic_df


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Category  
0        0         A/5 21171   7.2500   NaN        S  Jóvenes  
1        0          PC 17599  71.2833   C85        C  Adultos  
2        0  STON/O2. 3101282   7.9250   NaN        S  Jóvenes  
3        0            113803  53.1000  C123        S  Adultos  
4        0            373450   8.0500   NaN        S  Adultos  
..     ...               ...      ...   ...      ...      ...  
886      0            211536  13.0000   NaN        S  Jóvenes  
887      0            112053  30.0000   B42        S  Jóvenes  
888      2        W./C. 6607  23.4500   NaN        S  SinDato  
889      0            111369  30.0000  C148        C  Jóvenes  
890      0            370376   7.7500   NaN        Q  Adultos  

[891 rows x 13 columns]

* Crear un nuevo dataframe que contenga la edad y las categorías.

In [30]:
print(titanic_df.dtypes)

PassengerId       int64
Survived          int64
Pclass            int64
Name             object
Sex              object
Age             float64
SibSp             int64
Parch             int64
Ticket           object
Fare            float64
Cabin            object
Embarked         object
Category       category
dtype: object


In [5]:
new_titanic_df = titanic_df[['Age', 'Category']]
new_titanic_df

Age Category
0    22.0  Jóvenes
1    38.0  Adultos
2    26.0  Jóvenes
3    35.0  Adultos
4    35.0  Adultos
..    ...      ...
886  27.0  Jóvenes
887  19.0  Jóvenes
888   NaN  SinDato
889  26.0  Jóvenes
890  32.0  Adultos

[891 rows x 2 columns]

* A partir de la dataframe anterior obtener que porcentaje de pasajeros fueron Infantes,Niños,Adolescentes,Jóvenes,Adultos, Adulto Mayor y SinDato.

In [6]:
conteo_categorias = new_titanic_df['Category'].value_counts()
porcentaje_categorias = ((conteo_categorias / len(new_titanic_df)) * 100).round(2)

porcentaje_categorias

Category
Adultos         33.33
Jóvenes         30.30
SinDato         19.87
Adolescentes     7.86
Infantes         4.94
Niños            2.81
Adulto Mayor     0.90
Name: count, dtype: float64

* Crear un nuevo dataframe que contenga las categorias, conteo, porcentaje, como se muestra a continuación.

Categoria  Conteo  Porcentaje %
0          1      20            20
1          2      30            30
2          3      10            10
3          4      40            40

In [7]:
data = {
    'Categoria':[1,2,3,4],
    'Conteo':[20,30,10,40],
    'Porcentaje %':[20,30,10,40]
}

nuevo_df = pd.DataFrame(data)
nuevo_df

Categoria  Conteo  Porcentaje %
0          1      20            20
1          2      30            30
2          3      10            10
3          4      40            40

**Pregunta 11:** Dado el dataset titanic:

* Obtener la cantidad de hombres y mujeres por clase.
* Crear un nuevo dataframe con los datos obtenidos donde la clase será el indice.


In [ ]:
# Ejemplo del dataframe final


Hombres  \
Clase                                                 
Primera Clase  Cantidad de hombres en primera clase   
Segunda Clase  Cantidad de hombres en segunda clase   
Tercera Clase  Cantidad de hombres en tercera clase   

                                            Mujeres  
Clase                                                
Primera Clase  Cantidad de Mujeres en primera clase  
Segunda Clase  Cantidad de Mujeres en segunda clase  
Tercera Clase  Cantidad de Mujeres en tercera clase

In [31]:
tabla = titanic_df.groupby(['Pclass', 'Sex']).size().unstack(fill_value=0)
tabla = tabla.rename(columns={'female':'Mujeres', 'male':'Hombres'})
tabla = tabla.reindex(columns=['Hombres', 'Mujeres'])
tabla.index = tabla.index.astype(str)
tabla = tabla.rename(index={'1':'Primera Clase', '2':'Segunda Clase', '3':'Tercera Clase'})
tabla

Sex            Hombres  Mujeres
Pclass                         
Primera Clase      122       94
Segunda Clase      108       76
Tercera Clase      347      144

**Pregunta 12:** Dado el dataset titanic:

* Se desea saber que porcentaje de sobrevivientes fueron mujeres(en general).
* Se desea saber que porcentaje de sobrevivientes fueron infantes-niños(0-12 años).


In [53]:
mujeres_supervivientes=(titanic_df[(titanic_df['Survived']==1) & (titanic_df['Sex']=='female')]).shape[0]
porcentaje=(mujeres_supervivientes/len(titanic_df))*100
porcentaje_redondeado = round(porcentaje)
print('El porcentaje de mujeres que sobrevivieron fue del:', porcentaje_redondeado, '%')

El porcentaje de mujeres que sobrevivieron fue del: 26 %


In [54]:
infantes_niños_supervivientes = (titanic_df[(titanic_df['Survived']==1) & (titanic_df['Age'].between(0,12, inclusive='both'))]).shape[0]
porcentaje_infantes_niños_supervivientes = (infantes_niños_supervivientes/len(titanic_df))*100
porce_redondeado_inf_niños_supervi = round(porcentaje_infantes_niños_supervivientes)
print('El porcentaje de infantes y niños que sobrevivieron fue del:', porce_redondeado_inf_niños_supervi, '%')

El porcentaje de infantes y niños que sobrevivieron fue del: 4 %


**Pregunta 13:** Dado el dataset titanic:

* Se desea saber la cantidad de personas sobrevivientes y no sobrevivientes en primera clase.
* A partir del resultado anterior se desea saber la cantidad de hombres y mujeres en ambos casos.

In [68]:
supervivientes_primera_cla = (titanic_df[(titanic_df['Pclass']==1) & (titanic_df['Survived']==1)]).shape[0]
no_supervivientes_primea_cla = (titanic_df[(titanic_df['Pclass']==1) & (titanic_df['Survived']==0)]).shape[0]
print('El numero de supervivientes de la primera clase fueron:', supervivientes_primera_cla,', y el de fallecidos de la misma clase fueron:', no_supervivientes_primea_cla)

El numero de supervivientes de la primera clase fueron: 136 , y el de fallecidos de la misma clase fueron: 80


In [82]:
supervivientes_primera_cla_hombres = (titanic_df[(titanic_df['Pclass']==1) & (titanic_df['Survived']==1) & (titanic_df['Sex']=='male')]).shape[0]
supervivientes_primera_cla_mujeres = supervivientes_primera_cla - supervivientes_primera_cla_hombres
print('De los supervivientes de la primera clase, hombres fueron:', supervivientes_primera_cla_hombres,', y mujeres fueron:', supervivientes_primera_cla_mujeres)

De los supervivientes de la primera clase, hombres fueron: 45 , y mujeres fueron: 91


**Pregunta 14:** Dado el dataset titanic:

* Se desea saber en que puerto abordaron más pasajeros de primera clase.

**Ejem:** Del total de 100 pasajeros en primera clase, en el "puerto 1" es donde abordaron más pasajeros de esta clase con un total de 60 pasejeros.

In [110]:
tabla1 = titanic_df.groupby(['Pclass', 'Embarked']).size().unstack(fill_value=0)
tabla1

Embarked   C   Q    S
Pclass               
1         85   2  127
2         17   3  164
3         66  72  353

Del puerto donde más pasajeros de primera clase embarcaron fue del puerto de Southampton con un total de 127 pasajeros.

**Pregunta 15:** Dado el dataset titanic:

* Se desea saber cuantos pasajeros de primera,segunda y tercera clase abordaron en el puerto de "Queenstown". 

In [111]:
tabla2 = titanic_df.groupby(['Pclass', 'Embarked']).size().unstack(fill_value=0)
tabla2.drop(columns={'C','S'}).sum()

Embarked
Q    77
dtype: int64

En el puerto de Queenstown embarcaron un total de 77 pasajeros